titus-ext/cassandra/src/test/java/com/netflix/titus/ext/cassandra/store/CassStoreHelperTest.java

```diff
@@ -0,0 +1,104 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.ext.cassandra.store;
+
+import java.util.Arrays;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Session;
+import org.cassandraunit.CassandraCQLUnit;
+import org.cassandraunit.dataset.cql.ClassPathCQLDataSet;
+import org.junit.Rule;
+import org.junit.Test;
+import rx.Observable;
+import rx.schedulers.Schedulers;
+
+import static com.datastax.driver.core.querybuilder.QueryBuilder.bindMarker;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.insertInto;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.select;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class CassStoreHelperTest {
+
+    private static final long STARTUP_TIMEOUT = 30_000L;
+
+    /**
+     * As Cassandra uses memory mapped files there are sometimes issues with virtual disks storing the project files.
+     * To solve this issue, we relocate the default embedded Cassandra folder to /var/tmp/embeddedCassandra.
+     */
+    private static final String CONFIGURATION_FILE_NAME = "relocated-cassandra.yaml";
+
+    @Rule
+    public CassandraCQLUnit cassandraCQLUnit = new CassandraCQLUnit(
+            new ClassPathCQLDataSet("tables.cql", "titus_integration_tests"),
+            CONFIGURATION_FILE_NAME,
+            STARTUP_TIMEOUT
+    );
+
+    @Test
+    public void paginatedResultSetCanBeFetchedAsync() {
+        int numberOfRecords = 500;
+        int numberOfPages = 10;
+        Session session = cassandraCQLUnit.getSession();
+        PreparedStatement insertStmt = session.prepare(insertInto("app_scale_jobs").values(
+                Arrays.asList("job_id", "ref_id"),
+                Arrays.asList(bindMarker(), bindMarker())
+        ));
+
+        for (int i = 0; i < numberOfRecords; i++) {
+            ResultSet resultSet = session.execute(insertStmt.bind("job-" + i, UUID.randomUUID()));
+            assertThat(resultSet.wasApplied()).isTrue();
+        }
+
+        PreparedStatement loadStmt = session.prepare(select("job_id", "ref_id").from("app_scale_jobs"));
+        Observable<ResultSet> results = new CassStoreHelper(session, Schedulers.io()).execute(loadStmt.bind()
+                // force pagination, and pages to be fetched on demand as the ResultSet is iterated on
+                .setFetchSize(numberOfRecords / numberOfPages));
+
+        results.doOnNext(rows -> rows.forEach(row ->
+                assertThat(row.getString(0)).startsWith("job-"))
+        ).toCompletable().await(1, TimeUnit.MINUTES);
+    }
+
+    @Test(expected = IllegalStateException.class /* the datastax driver complains that page fetching will cause a deadlock */ )
+    public void paginatedResultInCassandraThreadsThrowsException() {
```

Comment:
Minor - but potentially a common 'fetchAllPages' method that takes scheduler as a parameter could simplify both of these test case code

Reply:
> Minor - but potentially a common 'fetchAllPages' method that takes scheduler as a parameter could simplify both of these test case code

That would require pre-loading all rows in memory, which iterating on resultsets is trying to avoid. I can see a callback or stream/reactive based API for it (e.g. `walkAllPages(Consumer<T> processor)`), but I didn't see a need to overcomplicate ATM

Reply:
nvm, I see what mean. I agree code duplication on both test cases would be good, next time.